hashes/zkevm/src/keccak/mod.rs

```diff
@@ -568,6 +553,111 @@ impl<F: Field> KeccakCircuitConfig<F> {
+        36      0xcba        3           0         1        1  // 2nd round begin
+        37        0          2           0         0        0
+        38        0          1           0         0        0
+        39        0          0           1         0        0
+        40        0          0           1         0        0
+        41        0          0           1         0        0
+        42        0          0           1         0        0
+        43        0          0           1         0        0
+        */
+
+        meta.create_gate("word_value", |meta| {
+            let mut cb = BaseConstraintBuilder::new(MAX_DEGREE);
+            cb.require_equal(
+                "word value",
+                absorb_data.expr(),
+                meta.query_advice(keccak_table.word_value, Rotation::cur()),
+            );
+            cb.gate(q(q_enable, meta) * not::expr(q(q_padding, meta)))
+        });
+        meta.create_gate("bytes_left", |meta| {
+            let mut cb = BaseConstraintBuilder::new(MAX_DEGREE);
+            // bytes_left[i] == 0 in the first round
+            cb.condition(q_in_round(q_first, meta), |cb| {
+                cb.require_zero(
+                    "bytes_left needs to be zero on the first row",
+                    meta.query_advice(keccak_table.bytes_left, Rotation::cur()),
+                );
+            });
+            // is_paddings only be true when
+            cb.condition(q(q_padding, meta), |cb| {
+                for i in 0..num_rows_per_round {
+                    let bytes_left = meta.query_advice(keccak_table.bytes_left, Rotation(i as i32));
+                    let bytes_left_next =
+                        meta.query_advice(keccak_table.bytes_left, Rotation(i as i32 + 1));
+                    // real input only when !is_paddings[i] && i < NUM_BYTES_PER_WORD
+                    cb.require_equal(
+                        "if not padding, bytes_left decreases by 1, else, stay same",
+                        bytes_left_next,
+                        bytes_left.clone()
+                            - if i < 7 || i == num_rows_per_round - 1 {
+                                not::expr(
+                                    is_paddings[std::cmp::min(i, NUM_BYTES_PER_WORD - 1)].expr(),
+                                )
+                            } else {
+                                0.expr()
+                            },
+                    );
+                    cb.require_zero(
+                        "bytes_left should be 0 when padding",
+                        bytes_left * is_paddings[std::cmp::min(i, NUM_BYTES_PER_WORD - 1)].expr(),
+                    );
+                }
+            });
+            cb.condition(
+                (q_in_round(q_enable, meta)
+                    - q_in_round(q_padding, meta)
+                    - q_in_round(q_first, meta))
+                    * meta.query_advice(is_final, Rotation::next()),
```

Comment:
Is this the right condition? Just checking.

This is saying, in an absorb round which is not a dummy round, on the row before the `is_final`, enforce equality. `is_final` is enforced to only be the first row of a round, and it _should_ only be in the last round.

Reply:
This part has been rewritten.